browse-indexing/Leech.java

```diff
@@ -73,36 +75,43 @@ private boolean termExists(String t)
     //
     public BrowseEntry next() throws Exception
     {
-        if (tenum == null) {
-            if (leafReaders.isEmpty()) {
-                // Nothing left to do
-                return null;
-            }
+        for (;;) {
+            if (tenum == null) {
+                // Load the next reader in our list and position the term enum.
+
+                if (leafReaders.isEmpty()) {
+                    // Nothing left to do
+                    return null;
+                }
+
+                // Select our next LeafReader to work from
+                LeafReader ir = leafReaders.remove(0).reader();
+                Terms terms = ir.terms(this.field);
 
-            // Select our next LeafReader to work from
-            LeafReader ir = leafReaders.remove(0).reader();
-            Terms terms = ir.terms(this.field);
+                if (terms == null) {
+                    // Try the next reader
+                    continue;
+                }
 
-            if (terms == null) {
-                // Try the next reader
-                return next();
+                tenum = terms.iterator();
             }
 
-            tenum = terms.iterator();
-        }
+            BytesRef nextTerm = tenum.next();
+
+            if (nextTerm == null) {
+                // Exhausted this reader.  Try the next one.
+                tenum = null;
+                continue;
+            }
 
-        if (tenum.next() != null) {
-            String termText = tenum.term().utf8ToString();
+            String termText = nextTerm.utf8ToString();
 
             if (termExists(termText)) {
-                return new BrowseEntry(buildSortKey(termText), termText, termText) ;
+                return new BrowseEntry(buildSortKey(termText), termText, termText);
             } else {
-                return this.next();
+                // Try the next term
+                continue;
```

Comment:
Minor nitpick, but do you really need this else/continue since you're at the bottom of the for loop anyway? Maybe a comment at the bottom of the loop like `// Time to try the next term` would serve the same purpose. (Apologies if I'm missing something that makes this necessary!).

Reply:
Nope, quite right, it's the same either way.  I've amended the commit to leave the comment and drop the else/continue.